### PR TITLE
Transfer metadata rework

### PR DESCRIPTION
**Pull request by adecaro**
- Transfer metadata key does not include txid and metadata counter anymore. It is up to the validator to make sure these keys are used properly
- check that a transfer contains only metadata that has been validated
- extend interop integration tests to orion

Signed-off-by: Angelo De Caro <adc@zurich.ibm.com>

### CODE REVIEW

File: token/core/fabtoken/validator_transfer.go [hunk 1]
```diff
@@ -11,6 +11,8 @@ import (
 	"encoding/json"
 	"time"
 
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/vault/keys"
+
```
Comment (HagarMeir) [inline token/core/fabtoken/validator_transfer.go:15]: empty line

File: token/core/zkatdlog/crypto/validator/validator_transfer.go [hunk 2]
```diff
@@ -11,6 +11,8 @@ import (
 	"encoding/json"
 	"time"
 
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/vault/keys"
+
```
Comment (HagarMeir) [inline token/core/zkatdlog/crypto/validator/validator_transfer.go:15]: empty line

File: token/services/interop/htlc/transaction.go [hunk 3]
```diff
@@ -13,6 +13,8 @@ import (
 	"encoding/json"
 	"time"
 
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/vault/keys"
+
```
Comment (HagarMeir) [inline token/services/interop/htlc/transaction.go:17]: empty line

File: token/services/vault/keys/keys.go [hunk 4]
```diff
@@ -36,6 +37,8 @@ const (
 	IssueActionMetadata         = "iam"
 	TransferActionMetadata      = "tam"
 	TokenRequestMetadata        = "trmd"
+	// Move back to the interop/htlc package
```
Comment (HagarMeir) [inline token/services/vault/keys/keys.go:40]: What is this? a TODO?

File: integration/token/interop/views/htlc/exchange.go [hunk 5]
```diff
@@ -255,7 +255,14 @@ func (v *FastExchangeResponderView) Call(context view.Context) (interface{}, err
 	_, err = view2.AsInitiatorCall(context, v, func(context view.Context) (interface{}, error) {
 		// Scan for the pre-image, this will be the signal that the initiator has claimed responder's script
 		// TODO: fix timeout
```
Comment (HagarMeir) [inline integration/token/interop/views/htlc/exchange.go:257]: What is this TODO?

File: token/services/network/driver/network.go [hunk 6]
```diff
@@ -10,6 +10,9 @@ import (
 	"context"
 	"encoding/base64"
 	"fmt"
+	"time"
+
+	token2 "github.com/hyperledger-labs/fabric-token-sdk/token"
```
Comment (HagarMeir) [inline token/services/network/driver/network.go:16]: empty line

File: token/services/network/driver/network.go [hunk 7]
```diff
@@ -93,4 +96,6 @@ type Network interface {
 
 	// UnsubscribeTxStatusChanges unregisters a listener for transaction status changes for the passed id
 	UnsubscribeTxStatusChanges(id string, listener TxStatusChangeListener) error
+
+	ScanForKey(namespace string, startingTxID string, key string, timeout time.Duration, opts ...token2.ServiceOption) ([]byte, error)
```
Comment (HagarMeir) [inline token/services/network/driver/network.go:100]: add comment

File: token/services/network/fabric/network.go [hunk 8]
```diff
@@ -8,9 +8,15 @@ package fabric
...
 import (
 	"context"
+	"encoding/base64"
 	"encoding/json"
 	"strings"
 	"sync"
+	"time"
+
+	"go.uber.org/zap/zapcore"
+
```
Comment (HagarMeir) [inline token/services/network/fabric/network.go:18]: empty lines

File: token/services/network/fabric/network.go [hunk 9]
```diff
@@ -310,3 +317,63 @@ func (n *Network) SubscribeTxStatusChanges(txID string, listener driver.TxStatus
...
+			subKey, err := w.GetTransferMetadataKeyWithSubKey(k)
+			if err != nil {
+				continue
+			}
+			if subKey == key {
+				keyValue = v
+				return true, nil
+			}
+		}
+		logger.Debugf("scanning for preimage on [%s] not found", tx.TxID())
```
Comment (HagarMeir) [inline token/services/network/fabric/network.go:362]: this shouldn't say preimage....should be more general

File: token/services/network/network.go [hunk 10]
```diff
@@ -12,6 +12,9 @@ import (
 	"encoding/json"
 	"fmt"
 	"sync"
+	"time"
+
```
Comment (HagarMeir) [inline token/services/network/network.go:16]: empty line

File: token/services/network/network.go [hunk 11]
```diff
@@ -320,6 +323,10 @@ func (n *Network) UnsubscribeTxStatusChanges(id string, listener TxStatusChangeL
 	return n.n.UnsubscribeTxStatusChanges(id, listener)
 }
 
+func (n *Network) ScanForKey(namespace, startingTxID, key string, timeout time.Duration, opts ...token.ServiceOption) ([]byte, error) {
```
Comment (HagarMeir) [inline token/services/network/network.go:326]: add comment

File: token/services/network/orion/network.go [hunk 12]
```diff
@@ -9,6 +9,9 @@ package orion
 import (
 	"context"
 	"sync"
+	"time"
+
+	"github.com/hyperledger-labs/fabric-token-sdk/token"
```
Comment (HagarMeir) [inline token/services/network/orion/network.go:15]: empty line

File: token/services/vault/keys/keys.go (hunk 4 above)
Comment (adecaro) [reply token/services/vault/keys/keys.go:40]: fixed

File: token/core/fabtoken/validator.go [hunk 13]
```diff
@@ -245,12 +246,25 @@ func (v *Validator) VerifyTransfer(ledger driver.Ledger, inputTokens []*token2.T
...
 		if err := validator(ctx); err != nil {
 			return err
 		}
 	}
+
+	// Check that all metadata have been validated
+	counter := 0
+	for k, c := range ctx.MetadataCounter {
+		if c > 1 {
+			return errors.Errorf("metadata key [%s] appeared more than one time", k)
```
Comment (KElkhiyaoui) [inline token/core/fabtoken/validator.go:261]: more than once.

File: token/core/interop/htlc/validator.go [hunk 14]
```diff
@@ -49,3 +54,41 @@ func VerifyOwner(senderRawOwner []byte, outRawOwner []byte, now time.Time) (*htl
...
+func MetadataCheck(action Action, script *htlc.Script, op OperationType, sig []byte) (string, error) {
+	if op == Reclaim {
+		// No metadata in this case
+		return "", nil
+	}
+
+	// Unmarshal signature to ClaimSignature
+	claim := &htlc.ClaimSignature{}
+	if err := json.Unmarshal(sig, claim); err != nil {
+		return "", errors.Wrapf(err, "failed unmarshalling cliam signature [%s]", string(sig))
```
Comment (KElkhiyaoui) [inline token/core/interop/htlc/validator.go:68]: claim signature

File: token/services/interop/htlc/script.go [hunk 15]
```diff
@@ -26,6 +26,31 @@ type HashInfo struct {
...
+	if !i.HashFunc.Available() {
+		return errors.New("hash function not available")
+	}
+	if !i.HashEncoding.Available() {
+		return errors.New("encoding function not available")
+	}
+	return nil
+}
+
+// Image computes the image of the passer pre-image using the hash and encoding function of this struct
```
Comment (KElkhiyaoui) [inline token/services/interop/htlc/script.go:40]: passed pre-image

File: token/services/network/driver/network.go [hunk 16]
```diff
@@ -93,4 +94,8 @@ type Network interface {
 
 	// UnsubscribeTxStatusChanges unregisters a listener for transaction status changes for the passed id
 	UnsubscribeTxStatusChanges(id string, listener TxStatusChangeListener) error
+
+	// LookupKey searches for the passed key starting from the passed transaction id in the given namespace.
+	// The operation gets canceled if the passed timeout gets reached.
```
Comment (KElkhiyaoui) [inline token/services/network/driver/network.go:99]: if the passed timeout elapses

File: token/services/network/fabric/network.go (hunk 9 above)
Comment (KElkhiyaoui) [reply token/services/network/fabric/network.go:359]: indeed.

**Review by KElkhiyaoui — CHANGES_REQUESTED** [review]
Looks good to me.

File: integration/token/interop/views/htlc/exchange.go (hunk 5 above)
Comment (adecaro) [reply integration/token/interop/views/htlc/exchange.go:253]: removed

File: token/core/interop/htlc/validator.go (hunk 14 above)
Comment (KElkhiyaoui) [inline token/core/interop/htlc/validator.go]: unmarshalling claim signature

File: token/services/interop/htlc/script.go (hunk 15 above)
Comment (KElkhiyaoui) [inline token/services/interop/htlc/script.go]: the passed pre-image

**Review by KElkhiyaoui — APPROVED** [review]

**Pull request merged**